### ISSUE TITLE
Add quality gates support

### ISSUE THREAD

**Issue by reinoudk** [rewe-digital/terraform-provider-sonarcloud#76]
### Description

Quality gates are an important part of SonarCloud and it should be possible to manage them using Terraform.

#### Data sources

* sonarcloud_source_quality_gates

#### Resources

* sonarcloud_quality_gate
* sonarcloud_quality_gate_selection

#### Data Sources

```tf
data "sonarcloud_quality_gates" "all" {}
```

#### Resources

```tf
# Note: this already exists
resource "sonarcloud_project" "example" {
  key        = "my-unique-project-key"
  name       = "My not-unique project name"
  visibility = "private"
}

resource "sonarcloud_quality_gate" "example" {
  name       = "Example Quality Gate"
  is_default = false

  condition {
    metric = "blocker_violations"
    op     = "LT"
    error  = 10
  }

  condition {
    metric = "complexity"
    op     = "LT"
    error  = 20
  }
}

resource "sonarcloud_quality_gate_selection" "example" {
  gate_id     = sonarcloud_quality_gate.example.id
  project_key = sonarcloud_project.example.key
}

```

### Notes

I'm not sure if the `condition{}` nested blocks are still supported like this, I'll have to check.

I'm also not sure I like the name of the `sonarcloud_quality_gate_selection`  resource. Normally this would be something like a `binding`, but SonarCloud's api has the `select` and `deselect` methods..

**Comment by Bhutania** [comment]
I was thinking of this layout for Conditions, as it mirrors the schema I have defined. How would your example's schema work?

```
resource "sonarcloud_quality_gate" "example" {
  name       = "Example Quality Gate"
  is_default = false

  conditions = [
    condition {
      metric = "blocker_violations"
      op     = "LT"
      error  = 10
    },
    condition {
      metric = "complexity"
      op     = "LT"
      error  = 20
    }
  ]
}
```

I went ahead and created the draft PR, even though, I still have to do a lot of testing. Particularly for the update requests. I anticipate there being some issues with my current implementation.

**Comment by reinoudk** [comment]
I was hoping it would be possible to define this in the schema similar to using `schema.TypeSet`, as done [here](https://github.com/hashicorp/terraform-provider-helm/blob/main/helm/resource_release.go#L138) in the Helm provider. See [this example](https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release#example-usage---chart-repository) for the usage.

But I have to check if this is also supported by the Terraform Plugin Framework. IMHO the nested blocks are a bit easier to write as a user.

**Comment by Bhutania** [comment]
Looks like `schema.typeSet` is possible. I'll rework the resource to use it, since it looks a lot easier to maintain.
https://www.terraform.io/plugin/sdkv2/schemas/schema-types#typeset

**Comment by Bhutania** [comment]
> Looks like `schema.typeSet` is possible. I'll rework the resource to use it, since it looks a lot easier to maintain. https://www.terraform.io/plugin/sdkv2/schemas/schema-types#typeset

Actually on second glance it looks like the only possible schema types are: singleNestedAttribute, listNestedAttribute(What is being used) and mapNestedAttribute. I got confused by the plugin framework documentation and regular plugin documentation.
https://www.terraform.io/plugin/framework/schemas#attributes-1

**Comment by reinoudk** [comment]
You're right. It might be possible using the `Type` instead of `Attributes`, but I don't think it's worth the hassle.

**Comment by Bhutania** [comment]
I'm steadily working through getting this resource working. I have run into an unexpected quirk of the Sonarcloud api. I cannot control whether quality gate conditions apply to overall code or new code. [I have created a post on the sonarcloud forums](https://community.sonarsource.com/t/new-code-vs-overall-code-for-quality-gate-conditions-created-using-sonarcloud-web-api/68843), so hopefully I have a fix/workaround soon!

**Comment by reinoudk** [comment]
As commented on the post: this is done by prefixing the permissions. You already have the the prefixed versions in the function that checks for allowed options, so I think you're good there. I had a lot of meetings today so haven't had the time to look at the code. I'll try and have a look tomorrow :slightly_smiling_face:.

**Comment by Bhutania** [comment]
Hey @reinoudk,

Thanks for the insight on the Sonarsource forums. I copied the list off a website, so I wasn't exactly comprehending some of the values I was writing down (whoops... 😅). I have gotten the quality gate resource into somewhat of a reasonable state. It passes the two acceptance tests that I wrote. I'm going to finish writing and debugging the selection and then write a few more tests for both resources so that I'm fully confident in all my functionality. (I haven't tested is_default for the quality gate resource).

Feel free to start poking around and point out any other bugs that you find, I'll get right on fixing them!

**Comment by Bhutania** [comment]
Hey @reinoudk,

Hope you are doing well. I encountered an error with the go-sonarcloud repo, and I have opened an issue there.

I would like to get your opinion on another matter. I have been creating the `sonarcloud_quality_gate_selection` resource with the schema that you defined above in mind:

```
resource "sonarcloud_quality_gate_selection" "example" {
  gate_id     = sonarcloud_quality_gate.example.id
  project_key = sonarcloud_project.example.key
}
```
This works great if there is a single quality gate selected for a single project. However there are cases where a single quality gate resource could be selected for multiple projects. While this could be accounted for in the current implementation like this:
```
resource "sonarcloud_quality_gate_selection" "example_1" {
  gate_id        = sonarcloud_quality_gate.example.id
  project_key = sonarcloud_project.example_1.key
}
resource "sonarcloud_quality_gate_selection" "example_2" {
  gate_id        = sonarcloud_quality_gate.example.id
  project_key = sonarcloud_project.example_2.key
}
```
What I am thinking about reworking the schema into is something like follows:
```
resource "sonarcloud_quality_gate_selection" "example" {
  gate_id          = sonarcloud_quality_gate.example.id
  project_keys = [
        sonarcloud_project.example_1.key
        sonarcloud_project.example_2.key
  ]
}
```
This sort of implementation can help clean up resource clutter.

This implementation would require removing/moving the ID attribute defined here:
https://github.com/Bhutania/terraform-provider-sonarcloud/blob/a593ef48c08cffdd0af88780f863b30767c06c3e/sonarcloud/resource_quality_gate_selection.go#L19

And changing the `Type` of `project_key` into a `types.SetType` or set the `Attributes` to `tfsdk.SetNestedAttributes` (depending on whether we believe that keeping the `id` of the selection in the state is valuable).

Let me know what you think! 😄

**Comment by reinoudk** [comment]
Hi @Bhutania, I've been a bit ill the last few days so I haven't been able to look at the PR. I hope next days will be a bit better. 

I like the idea of supporting multiple projects keys, that would be a nice improvement!

**Comment by Bhutania** [comment]
Oh no ☹️, hope you feel better soon!

I have reworked the resource, let me know what you think when you are feeling better!

**Comment by reinoudk** [comment]
Quality gate support has been released with [v0.3.0](https://github.com/rewe-digital/terraform-provider-sonarcloud/releases/tag/v0.3.0)! Many thanks for the implementation!
